packages/langchain/src/openai/util.ts

```diff
@@ -0,0 +1,212 @@
+import {
+  AIMessage,
+  BaseMessage,
+  ChatMessage,
+  ToolMessage
+} from '@langchain/core/messages';
+import { ChatResult } from '@langchain/core/outputs';
+import { StructuredTool } from '@langchain/core/tools';
+import type {
+  OpenAiChatCompletionChoice,
+  OpenAiChatCompletionFunction,
+  OpenAiChatCompletionTool,
+  OpenAiChatMessage,
+  OpenAiChatCompletionOutput,
+  OpenAiChatCompletionParameters
+} from '@sap-ai-sdk/foundation-models';
+import { zodToJsonSchema } from 'zod-to-json-schema';
+import { OpenAiChatClient } from './chat.js';
+import { OpenAIChatCallOptions } from './types.js';
+
+/**
+ * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionFunction}.
+ * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
+ * @returns The OpenAI Chat Completion Function.
+ * @internal
+ */
+export function mapToolToOpenAIFunction(
+  tool: StructuredTool
+): OpenAiChatCompletionFunction {
+  return {
+    name: tool.name,
+    description: tool.description,
+    parameters: zodToJsonSchema(tool.schema)
+  };
+}
+
+/**
+ * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionTool}.
+ * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
+ * @returns The OpenAI Chat Completion Tool.
+ * @internal
+ */
+export function mapToolToOpenAITool(
+  tool: StructuredTool
+): OpenAiChatCompletionTool {
+  return {
+    type: 'function',
+    function: {
+      name: tool.name,
+      description: tool.description,
+      parameters: zodToJsonSchema(tool.schema)
+    }
+  };
+}
+
+/**
+ * Maps a {@link BaseMessage} to OpenAI's Message Role.
+ * @param message - The message to map.
+ * @returns The OpenAI Message Role.
+ * @internal
+ */
+export function mapBaseMessageToRole(
+  message: BaseMessage
+): OpenAiChatMessage['role'] {
+  switch (message._getType()) {
+    case 'ai':
+      return 'assistant';
+    case 'human':
+      return 'user';
+    case 'system':
+      return 'system';
+    case 'function':
+      return 'function';
+    case 'tool':
+      return 'tool';
+    case 'generic':
+      return (message as ChatMessage).role as OpenAiChatMessage['role'];
```

Comment:
[q] If a message type is generic, the role field still aligns with the recognized roles in OpenAiChatMessage?

Reply:
I don't think so, I saw llm-commons doing this and was wondering the same thing.
I was searching through langchains API docs but didn't find any class which has the type "generic".

Reply:
I'll check if LLM Access has some special message type for their API, if they do then I should probably just change the type ..., otherwise I will remove this special case handling.